addon/corpus.js

```diff
@@ -1,106 +1,231 @@
-async function freezeAllPages() {
-    // Grey out Freeze button:
-    document.getElementById('freeze').disabled = true;
+let gUrls;  // Array of [filename, url] to freeze.
+let gUrlIndex;  // Pointer to current url in gUrls.
+let gFreezeOptions; // Freeze-dry options from UI.
+let gTimeout; // Load+Freeze timeout from UI.
+
+async function freezeAllPages(event) {
+    event.preventDefault();
+
+    // Check form validity.
+    if (!(
+        document.getElementById('wait').validity.valid &&
+        document.getElementById('timeout').validity.valid
+    )) {
+        return;
+    }
+
+    initialiseFromForm();
+    emptyElement(document.getElementById('status'));
 
-    // Clear error field:
-    const errorField = document.getElementById('errors');
-    while (errorField.firstChild) {
-        errorField.removeChild(errorField.firstChild);
+    // We need at least one url.
+    if (!gUrls.length) {
+        let li = document.createElement('li');
+        li.classList.add('error');
+        li.appendChild(document.createTextNode('no pages to download'));
+        document.getElementById('status').appendChild(li);
+        return;
     }
 
-    // Make freezing window, and set its size. The blank page acts as a
+    document.getElementById('freeze').disabled = true;
+    let windowId = 0;
+
+    // Listen for tab events before we start creating tabs; this avoids race
+    // conditions between creating tabs and waiting for loading to complete.
+    browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
+        if (
+            tab.windowId !== windowId ||
+            tab.url === 'about:blank' ||
+            tab.status !== 'complete'
+        ) {
+            return;
+        }
+
+        if (tab.url.startsWith('moz-extension://')) {
+            // The blank placeholder page has loaded, start the freezing process.
+            document.dispatchEvent(new CustomEvent(
+                'fathom:next',
+                { detail: windowId }
+            ));
+
+        } else {
+            // Tab that needs to be frozen has loaded.
+
+            // Setup the timeout; this covers both the page load and freeze time.
+            let timer = setTimeout(() => {
+                // Timeout.
+                console.error(tab.url, 'timeout');
+                clearTimeout(timer);
+                setCurrentStatus('timeout', true);
+
+                // Close the tab and process the next url.
+                browser.tabs.remove(tab.id)
+                    .then(() => {
+                        document.dispatchEvent(new CustomEvent(
+                            'fathom:next',
+                            { detail: windowId }
+                        ));
+                    })
+            }, gTimeout * 1000);
```

Comment:
`6608	http://www.alycealexandra.com/thermomix/2017/3/14/ellens-top-10-thermomix-vegan-recipes-but-actually-15` causes 61.0.1 to spinny-wheel for longer than the timeout+freezedelay, at least on my first try. The status says "loading" while it does. I will try again and make a note here. (My choice of line here is fairly arbitrary.)